extends CharacterBody3D

@export var SPEED = 2.0
@export var FOLLOW_DISTANCE = 20
@export var health : ComponentHealth

var movement_target_position: Vector3 = Vector3(-3.0,0.0,2.0)

@onready var movement_target_poll_rate : Timer = $movement_target_poll_rate
@onready var navigation_agent: NavigationAgent3D = $Marker3D/NavigationAgent3D

func _ready():
	# These values need to be adjusted for the actor's speed
	# and the navigation layout.
	navigation_agent.path_desired_distance = 0.5
	navigation_agent.target_desired_distance = 0.5

	# Make sure to not await during _ready.
	actor_setup.call_deferred()

func actor_setup():
	# Wait for the first physics frame so the NavigationServer can sync.
	await get_tree().physics_frame

	player.player_position_update.connect(set_movement_target)
	health.died.connect(died)

	# Now that the navigation map is no longer empty, set the movement target.
	set_movement_target(movement_target_position)

func set_movement_target(movement_target: Vector3):
	if movement_target_poll_rate.is_stopped():
		navigation_agent.set_target_position(movement_target)
		movement_target_poll_rate.start()


func _physics_process(delta: float) -> void:
	if navigation_agent.is_navigation_finished():
		return
		
	# Add the gravity.
	if not is_on_floor():
		velocity += get_gravity() * delta

	var current_agent_position: Vector3 = global_position
	var next_path_position: Vector3 = navigation_agent.get_next_path_position()

	velocity = current_agent_position.direction_to(next_path_position) * SPEED 
	move_and_slide()

func died(_killer):
	queue_free()
